restapi/src/main/java/io/stargate/web/docsapi/service/query/search/resolver/impl/SubDocumentsResolver.java

```diff
@@ -0,0 +1,138 @@
+ * parameters:
+ *
+ * <ol>
+ *   <li>1. <code>false</code> - there is no split by the keys on the sub-path, effectively the
+ *       whole document on the sub-path is returned
+ *   <li>2. <code>true</code> - there is extra split by the key values in the doc on the sub-path,
+ *       effectively returns document per key
+ * </ol>
+ */
+public class SubDocumentsResolver implements DocumentsResolver {
+
+  private final Expression<FilterExpression> expression;
+
+  private final AbstractSearchQueryBuilder queryBuilder;
+
+  private final ExecutionContext context;
+
+  private final int keyDepth;
+
+  public SubDocumentsResolver(
+      Expression<FilterExpression> expression,
+      String documentId,
+      List<String> subDocumentPath,
+      ExecutionContext context) {
+    this(expression, documentId, subDocumentPath, context, false);
+  }
+
+  public SubDocumentsResolver(
+      Expression<FilterExpression> expression,
+      String documentId,
+      List<String> subDocumentPath,
+      ExecutionContext context,
+      boolean splitOnSubPathKeys) {
+    this.expression = expression;
+    this.context = createContext(context, subDocumentPath);
+    this.queryBuilder = new SubDocumentSearchQueryBuilder(documentId, subDocumentPath);
+    // key depth explained:
+    //  - one extra for the document id
+    //  - one extra on size if we need to split based on keys ids after sub-document path
+    this.keyDepth = subDocumentPath.size() + (splitOnSubPathKeys ? 2 : 1);
+  }
+
+  @Override
+  public Flowable<RawDocument> getDocuments(
+      QueryExecutor queryExecutor,
+      DocsApiConfiguration configuration,
+      String keyspace,
+      String collection,
+      Paginator paginator) {
+    String[] columns = QueryConstants.ALL_COLUMNS_NAMES.apply(configuration.getMaxDepth());
+
+    // prepare the query
+    return RxUtils.singleFromFuture(
+            () -> {
+              DataStore dataStore = queryExecutor.getDataStore();
+              BuiltQuery<? extends BoundQuery> query =
+                  queryBuilder.buildQuery(dataStore::queryBuilder, keyspace, collection, columns);
+              return dataStore.prepare(query);
+            })
+
+        // cache the prepared
+        .cache()
+        .flatMapPublisher(
+            prepared -> {
+              // bind (no values needed)
+              BoundQuery query = prepared.bind();
+
+              // execute by respecting the paging state
+              return queryExecutor.queryDocs(
+                  keyDepth,
+                  query,
+                  configuration.getSearchPageSize(),
+                  paginator.getCurrentDbPageState(),
+                  context);
+            })
+        .filter(
+            document -> {
+              Map<String, EvalRule<FilterExpression>> rules = EvalEngine.booleanRules();
+              rules.put(FilterExpression.EXPR_TYPE, new RawDocumentEvalRule(document));
+              return EvalEngine.evaluate(expression, rules);
```

Comment:
Yea, this was intentionally changed:

1. To support multi field search (yes the `DocumentSearchService` does support it in the current state, we block from the outisde)
2. To be able to resolve ORs right away or actually any kind of the expression (`$not` would be also supported right away)

If you think about this, it's the same cost as loading the whole document, imo if search has same latency as the get, we should be fine, right. And we would be adding support for any kind of expression.

Furthermore, if there is sub-path, loading is only going for that sub-path, thus we never load more than we would actually need. 

Plus, since this is all in the same partition, loading all once could be even more performant that going for the 2 network hops to get first the row to evaluate and then the needed fields/structure.

We can discuss this, but I am thinking there are multiple benefits without actually degrading anything here..

Reply:
> Furthermore, if there is sub-path, loading is only going for that sub-path, thus we never load more than we would actually need.

Good point. I missed it :facepalm:

Reply:
Current approach LGTM. Resolving this comment.